### ISSUE TITLE
Building toolchain for Windows on Linux32

### ISSUE THREAD

**Issue by calvinatintel** [foss-for-synopsys-dwc-arc-processors/toolchain#56]
Hello,

I'm getting error while building the toolchains for Windows on Linux 32 host.

``` bash
cd toolchain
patch -p1 < windows-installer/  build-elf32_windows.patch
patching file build-elf32.sh
Hunk #1 succeeded at 206 (offset 1 line).
Hunk #2 succeeded at 222 (offset 1 line).
./build-all.sh --cpu arc700 --no-uclibc --no-sim --no-pdf --rel-rpaths --config-extra '--enable-sim-endian=no --with-python=no LDFLAGS=-static' --no-unisrc --no-external-download --no-auto-pull --no-auto-checkout
Checking out GIT trees ...
Will not download external dependencies
START ELF32: Mon Apr 13 00:11:45 UTC 2015
Installing in /var/jenkins/workspace/ARC_Toolchain-Windows/INSTALL
Configuring tools ...
  finished configuring tools
Building tools ...
ERROR: tools build failed.
ERROR: arc-elf32- tool chain build failed.
```

`logs-4.8/elf32-build-2015-04-13-0011.log` showed these errors:
http://pastebin.com/Eeb1L7Mq

```
configure: error: in `/var/jenkins/workspace/ARC_Toolchain-Windows/bd-4.8-elf32/intl':
configure: error: C compiler cannot create executables
See `config.log' for more details.
make: *** [configure-intl] Error 1
make: *** Waiting for unfinished jobs....
```

```
checking for library containing strerror... configure: error: Link tests are not allowed after GCC_NO_EXECUTABLES.
make: *** [configure-libiberty] Error 1
```

and `bd-4.8-elf32/intl/config.log` showed these errors:
http://pastebin.com/weiN7YVe

```
gcc version 4.8.2 (Ubuntu 4.8.2-19ubuntu1) 
configure:2958: $? = 0
configure:2947: gcc -V >&5
gcc: error: unrecognized command line option '-V'
gcc: fatal error: no input files
compilation terminated.
configure:2958: $? = 4
configure:2947: gcc -qversion >&5
gcc: error: unrecognized command line option '-qversion'
gcc: fatal error: no input files
compilation terminated.
configure:2958: $? = 4
configure:2978: checking for C compiler default output file name
configure:3000: gcc -g -O2 -D__USE_MINGW_ACCESS  -static-libstdc++ -static-libgcc -static -Wl,--stack,12582912 conftest.c  >&5
/usr/bin/ld: unrecognized option '--stack'
/usr/bin/ld: use the --help option for usage information
collect2: error: ld returned 1 exit status
configure:3004: $? = 1
configure:3041: result: 
configure: failed program was:
| /* confdefs.h */
| #define PACKAGE_NAME ""
| #define PACKAGE_TARNAME ""
| #define PACKAGE_VERSION ""
| #define PACKAGE_STRING ""
| #define PACKAGE_BUGREPORT ""
| #define PACKAGE_URL ""
| /* end confdefs.h.  */
| 
| int
| main ()
| {
| 
|   ;
|   return 0;
| }
configure:3047: error: in `/var/jenkins/workspace/ARC_Toolchain-Windows/bd-4.8-elf32/intl':
configure:3051: error: C compiler cannot create executables
See `config.log' for more details.
```

I've also looked at `bd-4.8-elf32/config.log` and found these:
http://pastebin.com/10J9Wp6d

```
configure:6039: checking for version 0.17.0 of CLooG
configure:6056: gcc -c -g -O2 -DCLOOG_INT_GMP   -I$$r/$(HOST_SUBDIR)/gmp -I$$s/gmp -I$$r/$(HOST_SUBDIR)/mpfr/src -I$$s/mpfr/src -I$$s/mpc/src   conftest.c >&5
conftest.c:10:27: fatal error: cloog/version.h: No such file or directory
 #include "cloog/version.h"
                           ^
compilation terminated.
configure:6056: $? = 1
configure: failed program was:
| /* confdefs.h */
| #define PACKAGE_NAME ""
| #define PACKAGE_TARNAME ""
| #define PACKAGE_VERSION ""
| #define PACKAGE_STRING ""
| #define PACKAGE_BUGREPORT ""
| #define PACKAGE_URL ""
| #define LT_OBJDIR ".libs/"
| /* end confdefs.h.  */
| #include "cloog/version.h"
| int
| main ()
| {
| #if CLOOG_VERSION_MAJOR != 0     || CLOOG_VERSION_MINOR != 17     || CLOOG_VERSION_REVISION < 0
|     choke me
|    #endif
|   ;
|   return 0;
| }
configure:6062: result: no
configure:6081: checking for version 0.18.0 of CLooG
configure:6098: gcc -c -g -O2 -DCLOOG_INT_GMP   -I$$r/$(HOST_SUBDIR)/gmp -I$$s/gmp -I$$r/$(HOST_SUBDIR)/mpfr/src -I$$s/mpfr/src -I$$s/mpc/src   conftest.c >&5
conftest.c:10:27: fatal error: cloog/version.h: No such file or directory
 #include "cloog/version.h"
                           ^
compilation terminated.
configure:6098: $? = 1
configure: failed program was:
| /* confdefs.h */
| #define PACKAGE_NAME ""
| #define PACKAGE_TARNAME ""
| #define PACKAGE_VERSION ""
| #define PACKAGE_STRING ""
| #define PACKAGE_BUGREPORT ""
| #define PACKAGE_URL ""
| #define LT_OBJDIR ".libs/"
| /* end confdefs.h.  */
| #include "cloog/version.h"
| int
| main ()
| {
| #if CLOOG_VERSION_MAJOR != 0     || CLOOG_VERSION_MINOR != 18     || CLOOG_VERSION_REVISION < 0
|     choke me
|    #endif
|   ;
|   return 0;
| }
configure:6104: result: no
```

Finally, there were also some errors in `bd-4.8-elf32/libiberty/config.log`:
http://pastebin.com/a5uvdabp

```
gcc: error: unrecognized command line option '-V'
gcc: fatal error: no input files
compilation terminated.
configure:3112: $? = 4
configure:3101: gcc -qversion >&5
gcc: error: unrecognized command line option '-qversion'
gcc: fatal error: no input files
compilation terminated.
configure:3112: $? = 4
configure:3128: gcc -o conftest -g -O2 -D__USE_MINGW_ACCESS  -static-libstdc++ -static-libgcc -static -Wl,--stack,12582912 conftest.c  >&5
/usr/bin/ld: unrecognized option '--stack'
/usr/bin/ld: use the --help option for usage information
collect2: error: ld returned 1 exit status
```

```
configure:3666: gcc -c -g -O2 -D__USE_MINGW_ACCESS  conftest.c >&5
conftest.c:15:3: warning: left shift count >= width of type [enabled by default]
   int off_t_is_large[(LARGE_OFF_T % 2147483629 == 721
   ^
conftest.c:15:3: warning: left shift count >= width of type [enabled by default]
conftest.c:16:10: warning: left shift count >= width of type [enabled by default]
          && LARGE_OFF_T % 2147483647 == 1)
          ^
conftest.c:16:10: warning: left shift count >= width of type [enabled by default]
conftest.c:15:7: error: variably modified 'off_t_is_large' at file scope
   int off_t_is_large[(LARGE_OFF_T % 2147483629 == 721
       ^
configure:3666: $? = 1
configure: failed program was:
| /* confdefs.h */
| #define PACKAGE_NAME ""
| #define PACKAGE_TARNAME ""
| #define PACKAGE_VERSION ""
| #define PACKAGE_STRING ""
| #define PACKAGE_BUGREPORT ""
| #define PACKAGE_URL ""
| /* end confdefs.h.  */
| #include <sys/types.h>
|  /* Check that off_t can represent 2**63 - 1 correctly.
|     We can't simply define LARGE_OFF_T to be 9223372036854775807,
|     since some C++ compilers masquerading as C compilers
|     incorrectly reject 9223372036854775807.  */
| #define LARGE_OFF_T (((off_t) 1 << 62) - 1 + ((off_t) 1 << 62))
|   int off_t_is_large[(LARGE_OFF_T % 2147483629 == 721
|              && LARGE_OFF_T % 2147483647 == 1)
|             ? 1 : -1];
| int
| main ()
| {
| 
|   ;
|   return 0;
| }
configure:3690: gcc -c -g -O2 -D__USE_MINGW_ACCESS  conftest.c >&5
configure:3690: $? = 0
configure:3698: result: 64
configure:3786: checking how to run the C preprocessor
configure:3817: gcc -E  conftest.c
configure:3817: $? = 0
configure:3831: gcc -E  conftest.c
conftest.c:10:28: fatal error: ac_nonexistent.h: No such file or directory
 #include <ac_nonexistent.h>
                            ^
compilation terminated.
configure:3831: $? = 1
configure: failed program was:
| /* confdefs.h */
| #define PACKAGE_NAME ""
| #define PACKAGE_TARNAME ""
| #define PACKAGE_VERSION ""
| #define PACKAGE_STRING ""
| #define PACKAGE_BUGREPORT ""
| #define PACKAGE_URL ""
| #define _FILE_OFFSET_BITS 64
| /* end confdefs.h.  */
| #include <ac_nonexistent.h>
configure:3856: result: gcc -E
configure:3876: gcc -E  conftest.c
configure:3876: $? = 0
configure:3890: gcc -E  conftest.c
conftest.c:10:28: fatal error: ac_nonexistent.h: No such file or directory
 #include <ac_nonexistent.h>
                            ^
compilation terminated.
configure:3890: $? = 1
configure: failed program was:
| /* confdefs.h */
| #define PACKAGE_NAME ""
| #define PACKAGE_TARNAME ""
| #define PACKAGE_VERSION ""
| #define PACKAGE_STRING ""
| #define PACKAGE_BUGREPORT ""
| #define PACKAGE_URL ""
| #define _FILE_OFFSET_BITS 64
| /* end confdefs.h.  */
| #include <ac_nonexistent.h>
```

and more.

Do you have a guide on how to cross compile for Windows?
Or can you spot what I'm doing wrong?

I'd appreciate your help.

Thank you,
Calvin

**Comment by anthony-kolesov** [comment]
Hi Calvin,

The guide is in https://github.com/foss-for-synopsys-dwc-arc-processors/toolchain/tree/arc-4.8-dev/windows-installer#notes-cross-compiling-for-windows and since you are using build-elf32_windows.patch I assume that you've seen it :) Building of toolchain for Windows is really a series of hacks applied one on top of another...

Main points:
1) First it is required to build ARC toolchain for Linux and add it to the PATH before building toolchain for Windows. Although to be honest, I never really tried to confirm myself that this is reqlly needed :)
2) build-elf32.sh should be modified to specify different --host value.
3) Instruction mentions manipulation with auto-build.h file... I already include one in this repository, although I cannot guarantee it will work for everybody.
4) In addition, build-elf32_windows.patch you are using also adds some extra hack with modification of the auto-host.h, that I also needed to get toolchain built.

There is also a file `Makefile.release` in this repository - I use it to generate our binary releases, and it automates this process, you can invoke it for Windows like:

```
make -f Makefile.release RELEASE=2014.12 ../release_output/.stamp_elf_le_windows_built
```

It will build toolchain for Linux as a byproduct, because it is required to build toolchain for Windows. 

Looking at your config.log files I have an impression that `configure` scripts have problems with detecting a proper MinGW compiler and instead uses native (for Linux) compiler:

```
configure:2546: checking host system type
configure:2559: result: i686-pc-mingw32
configure:2584: checking for i686-pc-mingw32-ar
configure:2611: result: i686-pc-mingw32-ar
configure:2676: checking for i686-pc-mingw32-ranlib
configure:2703: result: true
configure:2796: checking for i686-pc-mingw32-gcc
configure:2823: result: gcc
```

I think this happens because `--host` values should be `--host=i686-w64-mingw32` instead of current `--host=i686-pc-mingw32`. As far as I can tell, MingW on Ubuntu had `w64` in the triplet, while RHEL6 had `pc`, and I use RHEL myself to prepare releases. However I see that after RHEL updated MinGW GCC from 4.4 to 4.8 it now also has `w64` triplet, so it now unified with Ubuntu in this regard.

**Comment by calvinatintel** [comment]
**For anyone having trouble with Ubuntu 12, you must use Ubuntu 14.**

Hi Anthony,

With your help, I am _almost_ able to build the Windows.

At the end of the build, Makefile calls this:
`cp -rfdv /usr/i686-pc-mingw32/sys-root/mingw/bin/* $O/$(WIN_ELF_LE_DIR)/bin`

This fails because my build machine doesn't have `/usr/i686-pc-mingw32` directory despite having installed `mingw-w64` package. However, the toolchain still seems to work.

Do you know what that copy command is supposed to get from that directory?

Thank you!

**Comment by anthony-kolesov** [comment]
Hi Calvin,

That was also a reference to older mingw triplet (i686-pc-mingw32), so it also should be changed to i686-w64-mingw32. Here are files in this directory:

```
$ ls /usr/i686-w64-mingw32/sys-root/mingw/bin/
libgcc_s_sjlj-1.dll  libssp-0.dll  libstdc++-6.dll  libwinpthread-1.dll
```

I'm not recall if those files are just copied because they are in the sysroot and better safe than sorry, or because they are really needed, but libstdc++ definitely should be needed unless toolchain has been linked statically.

**Comment by calvinatintel** [comment]
Thank you as always for the prompt and accurate replies.

Do you know which package provides `/usr/i686-w64-mingw32/sys-root/mingw/bin/libstdc++-6.dll`?

[This search](http://packages.ubuntu.com/search?suite=utopic&arch=any&mode=filename&searchon=contents&keywords=libstdc%2B%2B-6.dll) thinks that [g++-mingw-w64-i686](http://packages.ubuntu.com/utopic/g%2B%2B-mingw-w64-i686) is the closest thing to it, but the file is in wrong directory. Did you have to compile and install mingw manually?

Also, do you know how I can compile the toolchain to link statically? Will this flag do? `--config-extra "LDFLAGS=-static"`

**Comment by anthony-kolesov** [comment]
Yeah, I see that on Ubuntu directory structure is different, than on RHEL. So you need to modify this `Makefile.release` to work on Ubuntu. Patches welcome :)

For static toolchain that option should be fine, at least that seems to work with our own prebuilt toolchain.

**Comment by calvinatintel** [comment]
`libgcc_s_sjlj-1.dll` and `libssp-0.dll` comes from [gcc-mingw-w64-i686](http://packages.ubuntu.com/utopic/gcc-mingw-w64-i686)
`libstdc++-6.dll` comes from [g++-mingw-w64-i686](http://packages.ubuntu.com/utopic/g%2B%2B-mingw-w64-i686)
`libwinpthread-1.dll` comes from [mingw-w64-i686-dev](http://packages.ubuntu.com/utopic/mingw-w64-i686-dev)

What do you suggest is the right way to patch this?
1. We could list those three packages as dependencies and copy from `/usr/lib/gcc/i686-w64-mingw32/4.9-win32/*.dll`. However, that directory contains more than just those four DLL's.
2. We could copy those four exact DLL's
3. There's a guy who provides [all the mingw DLL's](http://qemu.weilnetz.de/w32/dll/). If we know exactly which DLL's are required, we can just download from his site. Or, we can even just commit them into this repo and copy over.

**Comment by calvinatintel** [comment]
Also, which branches would you like me to submit the patch to?
I want to make sure it'll be backported to `arc-2014.12` release.

**Comment by anthony-kolesov** [comment]
Hi Calvin,

It looks like that copy of those files has been added as of precaution, since they are present in sysroot, but not because there was an actual need. I've checked some of our exe files in Windows release and as far as I see they are not dependant on libgcc, since linked statically. I will remove that file copying from our next release.

**Comment by anthony-kolesov** [comment]
Fixed in arc-dev as 4b6277136b2ec36b1861692eab18220ea77b156b
Fixed in arc-4.8-dev as c85b5c1ca2477d4289a91d17f815d31bc06b34c3

**Comment by calvinatintel** [comment]
Thank you for your wonderful support
